### PR TITLE
Including config file 

### PR DESCRIPTION
**Pull request by hjsuh94**
First pass on config file....ended up being quite a pain because of all the dependencies but I have a working version now. 

I don't really like this solution because it feels pythonic (I'm really just routing config "dictionaries" around into constructor arguments of every class that needs it), so let me know if there is a more elegant solution.

+@pangtao22 +@mpetersen94 for review. It is a working version, but feel free to comment on high-level stuff first if you think something should be different.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)

### CODE REVIEW

**Review by hjsuh94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)* status: 0 of 21 files reviewed, 2 unresolved discussions (waiting on @hjsuh94)

___
*[src/iiwa_plan_manager.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_1S31_Vkz3I0xesF:-MYr_1S31_Vkz3I0xesG:bxo6es4) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/30cf23e3e7c1ecdd952190795564d5e498ed1767/src/iiwa_plan_manager.h#L19)):*
> ```objc
> 
>  private:
>   const double control_period_seconds_;
> ```

I would like to preserve the const here, but this isn't possible because this variable is being initialized from a constructor argument (The YAML Node)
___
*[src/plans/task_space_trajectory_plan.cc, line 14 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_EEk-nu0a5uFXGAA:-MYr_EElDlAo-JKDsgq-:b-ka5oy9) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/src/plans/task_space_trajectory_plan.cc#L14)):*
> ```cpp
>   // TODO(terry-suh): Get this from a config file.
>   const drake::multibody::Frame<double> &frame_E =
>       plant_->GetFrameByName(config_["robot_ee_body_name"].as<std::string>());
> ```

Each plans will need access to the config because of things like this. YAML gets passed down by the following tree:

`iiwa_plan_manager` ------------------------------------------------------------------------------|
`iiwa_plan_manager_hardware_interface` -> `iiwa_plan_manager_system` ----|-> `iiwa_plan_factory` -> `plans`

File: src/iiwa_plan_manager.cc [hunk 1]
```diff
@@ -154,7 +159,8 @@ void IiwaPlanManager::HandleIiwaStatus(
   bool command_has_error;
   {
     std::lock_guard<std::mutex> lock(mutex_state_machine_);
-    command_has_error = state_machine_->CommandHasError(s, c);
+    command_has_error = state_machine_->CommandHasError(s, c,
+      config_["q_threshold"].as<double>());
```
Comment (hjsuh94) [inline src/iiwa_plan_manager.cc:163]: I didn't want to route the config files through all the state machine stuff, so I ended up just having a threshold as an argument to the error function.

File: src/iiwa_plan_manager.h [hunk 2]
```diff
@@ -1,41 +1,41 @@
...
 class IiwaPlanManager {
- public:
-  IiwaPlanManager(double control_period);
+public:
+  IiwaPlanManager(YAML::Node config);
   ~IiwaPlanManager();
   void Run();
 
- private:
-  const double control_period_seconds_;
```
Comment (hjsuh94) [inline src/iiwa_plan_manager.h:19]: same comment here as below. const is gone

**Review by pangtao22 — CHANGES_REQUESTED** [review]
Some high-level comments:
- Does it makes sense to create a structure with correctly typed fields, write a function to make the structure from the yaml node, and pass the structure around instead? This will get rid of the type casting every time a filed in the yaml node is accessed. 
- I don't think `*Plan` objects need access to the full configuration. `JointSpaceTrajectoryPlan` needs none, `TaskSpaceTrajectoryPlan` needs only the name of the frame the plan runs on. For instance, if we define TaskSpacePlan like this: 
```c++
class TaskSpaceTrajectoryPlan : public PlanBase {
public:
  TaskSpaceTrajectoryPlan(/*trajectories*/, const std::string& frame_name) {
    /*other stuff*/
   frame_id_ = plant_->GetFrameByName(frame_name).index();
  }
private:
  drake::multibody::FrameIndex frame_id_;
}
```
we can eliminate the dependency of `PlanBase` on the configuration. 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYrf0Py-CpAu5VZtuKR:b-7t77fd)* status: 0 of 27 files reviewed, 8 unresolved discussions (waiting on @hjsuh94 and @mpetersen94)

___
*[config/default.yaml, line 10 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrctXQ2oK46Zexw1XF:-MYrctXQ2oK46Zexw1XG:b-kinqkt) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/config/default.yaml#L10)):*
> ```YAML
> robot_sdf_path: "drake/manipulation/models/iiwa_description/iiwa7/iiwa7_no_collision.sdf"
> robot_baselink_name: "iiwa_link_0"
> robot_ee_body_name: "iiwa_link_5"
> ```

The frame for which we'd like to run a `TaskSpaceTrajectoryPlan` could change on the fly. For instance, we can move the EE in one plan and link 6 in another. (And I think the default EE is link 7?) A better place to send this is perhaps our new lcm type for plans?
___
*[src/iiwa_plan_manager.h, line 2 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_Uq6-As8E_O_KJ78:-MYr_Uq6-As8E_O_KJ79:bi34zcv) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/src/iiwa_plan_manager.h#L2)):*
> ```objc
> #pragma once
> #include "drake_lcmtypes/drake/lcmt_iiwa_status.hpp"
> ```

IIRC, the recommended order of includes is stl first, external libraries second and finally our own headers. 
___
*[src/iiwa_plan_manager.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_1S31_Vkz3I0xesF:-MYrdv2N7QpOk5KKHLE_:b-v61k7) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/30cf23e3e7c1ecdd952190795564d5e498ed1767/src/iiwa_plan_manager.h#L19)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

I would like to preserve the const here, but this isn't possible because this variable is being initialized from a constructor argument (The YAML Node)
</blockquote></details>

I think it's possible to initialize a const data member from constructor arguments. I'll give it a try. 
___
*[src/iiwa_plan_manager.cc, line 13 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYre6wnCxJnMNzc7lNZ:-MYre6wnCxJnMNzc7lN_:b-gcfhjr) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/iiwa_plan_manager.cc#L13)):*
> ```cpp
> using std::endl;
> 
> IiwaPlanManager::IiwaPlanManager(YAML::Node config) : config_(config) {
> ```

`config_(config)` can be replaced with `config_(std::move(config))` to avoid unnecessary copying of `config`.
___
*[src/plans/iiwa_plan_factory.cc, line 19 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYriZid2dbGTM9Y-jj5:-MYriZid2dbGTM9Y-jj6:b6fgly6) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/plans/iiwa_plan_factory.cc#L19)):*
> ```cpp
> using std::vector;
> 
> IiwaPlanFactory::IiwaPlanFactory(YAML::Node config_) {
> ```

IiwaPlanFactory can keep a constant reference to the configuration, which should be owned by `IiwaPlanManager`. This would get rid of the `YAML::Node config` argument in every method of this class. 
___
*[src/plans/task_space_trajectory_plan.cc, line 13 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrg_hjCOrp86VUcxN5:-MYrg_hk30i4lHmaYCFq:bzd1iwb) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/plans/task_space_trajectory_plan.cc#L13)):*
> ```cpp
>   plant_->SetPositions(plant_context_.get(), state.q);
>   // TODO(terry-suh): Get this from a config file.
>   const drake::multibody::Frame<double> &frame_E =
> ```

We can get this frame in the constructor and save its index as a data member of the class. Doing this avoids calling`GetFrameByName` at 200Hz, which has linear complexity in the number of frames in the MBP.

**Review by pangtao22 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)* status: 0 of 27 files reviewed, 8 unresolved discussions (waiting on @hjsuh94 and @mpetersen94)

___
*[src/iiwa_plan_manager.cc, line 163 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrY9So4jrqF65YYhTa-r2-163:-MYreqBKCiddLKVfFKms:b8lg1k) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/iiwa_plan_manager.cc#L163)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

I didn't want to route the config files through all the state machine stuff, so I ended up just having a threshold as an argument to the error function. 
</blockquote></details>

It might make sense for `state_machine` to keep a constant reference to the configuration as well, as both `StateMachine` and `IiwaPlanFactory` are owned by `IiwaPlanManager`.
___
*[src/plans/task_space_trajectory_plan.cc, line 14 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_EEk-nu0a5uFXGAA:-MYrjcpy4abVQdJCUMIg:b-9mux7i) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/src/plans/task_space_trajectory_plan.cc#L14)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

Each plans will need access to the config because of things like this. YAML gets passed down by the following tree:

`iiwa_plan_manager` ------------------------------------------------------------------------------|
`iiwa_plan_manager_hardware_interface` -> `iiwa_plan_manager_system` ----|-> `iiwa_plan_factory` -> `plans` 
</blockquote></details>

As mentioned in the high-level comments above, I think plans only need to store what's needed by the plan, not the entire configuration.

**Review by hjsuh94 — COMMENTED** [review]
Thanks Pang! Some questions below on the comments. 
1. I thought about this, but wonder if it's repetitive since config is a pretty good struct on its own. You're right about the type-casting part....would you recommend going with our own structs? 
2. I agree that most plan objects only need access to a small subset of config files. I was a bit worried about the scalability of different plans coming in the future, but I think you're right. 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYrko9k0U6oJU22iyj_:bcvfwuq)* status: 0 of 27 files reviewed, 8 unresolved discussions (waiting on @hjsuh94, @mpetersen94, and @pangtao22)

___
*[config/default.yaml, line 10 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrctXQ2oK46Zexw1XF:-MYrkS8dDBz4ynA7Ryyv:b-kifa36) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/config/default.yaml#L10)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

The frame for which we'd like to run a `TaskSpaceTrajectoryPlan` could change on the fly. For instance, we can move the EE in one plan and link 6 in another. (And I think the default EE is link 7?) A better place to send this is perhaps our new lcm type for plans?
</blockquote></details>

Huh....I've never thought about that and I agree that it would be a better way. I just ran through my head that this has no conflicts with your comment below about `GetFrameByName` too, since it can be initialized per-frame basis (not per-step basis). 
___
*[src/iiwa_plan_manager.h, line 2 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_Uq6-As8E_O_KJ78:-MYrkJgc1kcB8fNRgChV:bbemnyf) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/src/iiwa_plan_manager.h#L2)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

IIRC, the recommended order of includes is stl first, external libraries second and finally our own headers. 
</blockquote></details>

oops - I ran `clang-format-9` on the whole folder and it must have changed the ordering.
___
*[src/plans/iiwa_plan_factory.cc, line 19 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYriZid2dbGTM9Y-jj5:-MYrk43q9Wr6CLrRiV9C:bh2naaa) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/plans/iiwa_plan_factory.cc#L19)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

IiwaPlanFactory can keep a constant reference to the configuration, which should be owned by `IiwaPlanManager`. This would get rid of the `YAML::Node config` argument in every method of this class. 
</blockquote></details>

Ahh could you elaborate on this solution a bit more? Wouldn't it still require a pointer to IiwaPlanManager's `YAML::Node` as an argument?

**Review by mpetersen94 — APPROVED** [review]
I agree with Pang that passing the whole config to the plans feels excessive.  They definitely are getting more information than they need in the process.  Can we instead pull what is needed from the config and just pass that?  Might also help cut down on the type casting from the config.

> Reviewed 18 of 21 files at r1, 8 of 9 files at r2, 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYrpdpM4l8nQPjDGTGY:bf6cjc0)* status: all files reviewed, 8 unresolved discussions (waiting on @hjsuh94 and @pangtao22)

**Review by pangtao22 — COMMENTED** [review]
Here's a proposal for structuring the configuration based on the discussions so far:

- `IiwaPlanManager` owns the configuration as a yaml node (not a struct translated by us), `IiwaPlanFactory` has a constant pointer/reference to the yaml node.
- Every plan sub-class that requires a configuration has its own `XxxPlanConfiguration` struct. For example, `TaskSpaceTrajectoryPlan` may look like:
```c++
// in task_space_trajectory_plan.h
struct TaskSpaceTrajectoryPlanConfig {
  Eigen::Array3d position_feedback_gain;
  Eigen::Array3d orientation_feedback_gain;
  std::string frame_name;
}

class TaskSpaceTrajectoryPlan : public PlanBase {
public:
  TaskSpaceTrajectoryPlan(/*other stuff*/, TaskSpaceTrajectoryPlanConfig config) : 
    config_(std::move(config)) {/*constructor body*/}

private:
  const TaskSpaceTrajectoryPlanConfig config_;
  // other stuff.
}
```

The config struct is created inside the `IiwaPlanFactory::MakeTaskSpaceTrajectoryPlan` method. This should limit the type-casting to only the Factory class. 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYrqLYSCATFe6bXVyON:b-p8m6po)* status: all files reviewed, 8 unresolved discussions (waiting on @hjsuh94 and @pangtao22)

**Review by pangtao22 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)* status: all files reviewed, 8 unresolved discussions (waiting on @hjsuh94 and @pangtao22)

___
*[src/plans/iiwa_plan_factory.cc, line 19 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYriZid2dbGTM9Y-jj5:-MYruJ5_EpLlAM-NulwI:b-9bwrox) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/plans/iiwa_plan_factory.cc#L19)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

Ahh could you elaborate on this solution a bit more? Wouldn't it still require a pointer to IiwaPlanManager's `YAML::Node` as an argument?
</blockquote></details>

This is what I meant:
```c++
class IiwaPlanFactory {
public: 
  IiwaPlanFacotry(const YAML::Node config): config_(config) {/*constructor body*/}
  // other stuff
private:
  const YAML::Node& config_;  // defined as reference to avoid copying.
  // other stuff
}
```
The configuration therefore becomes available to methods of IiwaPlanFactory through the private `config_` field.

**Review by pangtao22 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)* status: all files reviewed, 8 unresolved discussions (waiting on @hjsuh94 and @pangtao22)

___
*[src/plans/iiwa_plan_factory.cc, line 19 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYriZid2dbGTM9Y-jj5:-MYrvJgYCKpD4lQC5oad:b7bpukg) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/plans/iiwa_plan_factory.cc#L19)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

This is what I meant:
```c++
class IiwaPlanFactory {
public: 
  IiwaPlanFacotry(const YAML::Node config): config_(config) {/*constructor body*/}
  // other stuff
private:
  const YAML::Node& config_;  // defined as reference to avoid copying.
  // other stuff
}
```
The configuration therefore becomes available to methods of IiwaPlanFactory through the private `config_` field.
</blockquote></details>

The constructor signature should be `IiwaPlanFacotry(const YAML::Node& config)`.

**Review by hjsuh94 — COMMENTED** [review]
One question here - why not just have everything in `TaskSpaceTrajectoryPlanConfig` as an argument to the constructor of `TaskSpaceTrajectoryPlan`? 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYu1Ncw6TIEU-Nkr_a-:bs6xulx)* status: all files reviewed, 8 unresolved discussions (waiting on @hjsuh94 and @pangtao22)

___
*[config/default.yaml, line 10 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrctXQ2oK46Zexw1XF:-MYu1W9B9QZamOhVQIfT:b-pxd7sm) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/config/default.yaml#L10)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

Huh....I've never thought about that and I agree that it would be a better way. I just ran through my head that this has no conflicts with your comment below about `GetFrameByName` too, since it can be initialized per-frame basis (not per-step basis). 
</blockquote></details>

I will leave this one in the config file for now and submit an issue in #15!
___
*[src/iiwa_plan_manager.cc, line 163 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrY9So4jrqF65YYhTa-r2-163:-MYu2AIMD3eQt-D9WzFn:bixh3m5) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/iiwa_plan_manager.cc#L163)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

It might make sense for `state_machine` to keep a constant reference to the configuration as well, as both `StateMachine` and `IiwaPlanFactory` are owned by `IiwaPlanManager`.
</blockquote></details>

Here we come across another slightly ambiguous decision. The implementation of `CommandHasError` is actually a method of `StateBase`, so we'll need to decide if the States need access to the config files as well. 

Are you on board with the following?
1. `state_machine` has a constant reference to the config file.
2. The `states` instances inside the state machine can only access config files through specific arguments in methods / constructor. Similar to the relation between factory and plans, this limits type casting to only state machines, but not states.

**Review by pangtao22 — COMMENTED** [review]
That would work too, but we'll need to declare them separately as private data members of `TaskSpaceTrajectoryPlan`, and the constructor of `TaskSpaceTrajectoryPlan` would have a longer signature.

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYu4GMr-llFpqanVc4g:b7jfzti)* status: all files reviewed, 6 unresolved discussions (waiting on @hjsuh94)

___
*[src/iiwa_plan_manager.cc, line 163 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrY9So4jrqF65YYhTa-r2-163:-MYu6_Lq5GGDmssmjFao:b5x6n1g) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/iiwa_plan_manager.cc#L163)):*
<details><summary><i>Previously, hjsuh94 (Hyung Ju Suh (Terry)) wrote…</i></summary><blockquote>

Here we come across another slightly ambiguous decision. The implementation of `CommandHasError` is actually a method of `StateBase`, so we'll need to decide if the States need access to the config files as well. 

Are you on board with the following?
1. `state_machine` has a constant reference to the config file.
2. The `states` instances inside the state machine can only access config files through specific arguments in methods / constructor. Similar to the relation between factory and plans, this limits type casting to only state machines, but not states. 
</blockquote></details>

That sounds good!

**Review by hjsuh94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)* status: 12 of 27 files reviewed, 4 unresolved discussions (waiting on @mpetersen94 and @pangtao22)

___
*[src/iiwa_plan_manager.h, line 19 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_1S31_Vkz3I0xesF:-MYuGV0v8j9u7ypEXxC6:b-s8cula) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/30cf23e3e7c1ecdd952190795564d5e498ed1767/src/iiwa_plan_manager.h#L19)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

I think it's possible to initialize a const data member from constructor arguments. I'll give it a try. 
</blockquote></details>

Ahh I was able to fix this by doing initializer lists from config! 
___
*[src/iiwa_plan_manager.cc, line 13 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYre6wnCxJnMNzc7lNZ:-MYuGa4o6QTaTZd2B0ks:b-896fix) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/iiwa_plan_manager.cc#L13)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

`config_(config)` can be replaced with `config_(std::move(config))` to avoid unnecessary copying of `config`.
</blockquote></details>

Done.
___
*[src/iiwa_plan_manager.cc, line 163 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrY9So4jrqF65YYhTa-r2-163:-MYuGbPC-EPf8Yr8nHoy:b-qf4mu1) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/iiwa_plan_manager.cc#L163)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

That sounds good! 
</blockquote></details>

Done. Now `state_machine_` will have the method `CommandHasError(s, c)`, but `state_base`will have the method `CommandHasError(s, c, q_threshold)`, and state machine is responsible for reading from config file and passing over the relevant parameters to the state. 
___
*[src/plans/iiwa_plan_factory.cc, line 19 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYriZid2dbGTM9Y-jj5:-MYuGqFF5koLoSekvKnO:bs7wumx) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/plans/iiwa_plan_factory.cc#L19)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

The constructor signature should be `IiwaPlanFacotry(const YAML::Node& config)`.
</blockquote></details>

Done. 
___
*[src/plans/task_space_trajectory_plan.cc, line 14 at r1](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYr_EEk-nu0a5uFXGAA:-MYuGrUJEC032wNLisG7:b-896fix) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/7bc34320173a720ae9f95a23cd158a292600077d/src/plans/task_space_trajectory_plan.cc#L14)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

As mentioned in the high-level comments above, I think plans only need to store what's needed by the plan, not the entire configuration.
</blockquote></details>

Done.
___
*[src/plans/task_space_trajectory_plan.cc, line 13 at r2](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYrg_hjCOrp86VUcxN5:-MYuGsB34O6_wu-PxKlq:b6r432q) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/627a03a7c60625a7f31ba3d3ba133de6d952c819/src/plans/task_space_trajectory_plan.cc#L13)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

We can get this frame in the constructor and save its index as a data member of the class. Doing this avoids calling`GetFrameByName` at 200Hz, which has linear complexity in the number of frames in the MBP. 
</blockquote></details>

Good catch! This has been fixed. Done.

**Review by hjsuh94 — COMMENTED** [review]
For now I will leave it in the constructor because there is really only one field, so making a struct felt a bit excessive :) I think each plan can make a choice on whether they want a struct or leave it in the constructor. 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYuHiZT5Xl7cMv3PS6R:b-e2mmmq)* status: 12 of 27 files reviewed, 4 unresolved discussions (waiting on @mpetersen94 and @pangtao22)

**Review by pangtao22 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)* status: 12 of 27 files reviewed, 8 unresolved discussions (waiting on @hjsuh94, @mpetersen94, and @pangtao22)

___
*[src/iiwa_plan_manager.h, line 26 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuF8CkCo-pDdrAxemx:-MYuF8CkCo-pDdrAxemy:bbatrlq) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/iiwa_plan_manager.h#L26)):*
> ```objc
>   std::unique_ptr<IiwaPlanFactory> plan_factory_;
> 
>   YAML::Node config_;
> ```

Is it possible to make it constant?
___
*[src/plan_manager_system/iiwa_plan_manager_hardware_interface.h, line 23 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuFhQN1s0EqQi6f_iN:-MYuFhQN1s0EqQi6f_iO:bp0yccf) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/plan_manager_system/iiwa_plan_manager_hardware_interface.h#L23)):*
> ```objc
>   std::unique_ptr<drake::systems::Diagram<double>> diagram_;
>   std::unique_ptr<drake::lcm::DrakeLcm> owned_lcm_;
>   YAML::Node config_;
> ```

`IiwaPlanManagerHardwareInterface` doesn't seem to use config outside the constructor. Maybe we don't need to save it as a data member here? It's probably better for the config to be owned by `IiwaPlanManagerSystem`?
___
*[src/plan_manager_system/iiwa_plan_manager_system.h, line 48 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuHBon30l4O6lnThlN:-MYuHBon30l4O6lnThlO:bnzmmd0) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/plan_manager_system/iiwa_plan_manager_system.h#L48)):*
> ```objc
>   drake::systems::OutputPortIndex output_port_iiwa_command_idx_;
> 
>   YAML::Node config_;
> ```

Is it possible to make it constant, similar to `IiwaPlanManager`?
___
*[src/plans/iiwa_plan_factory.cc, line 23 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuHiR-8UolhxEIBa6M:-MYuHiR-8UolhxEIBa6N:bom6m4a) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/plans/iiwa_plan_factory.cc#L23)):*
> ```cpp
>   auto parser = drake::multibody::Parser(plant_.get());
> 
>   std::string iiwa_path = config_["robot_sdf_path"].as<std::string>();
> ```

nit: this can be made a constant reference.

**Review by hjsuh94 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18)* status: 11 of 27 files reviewed, 7 unresolved discussions (waiting on @mpetersen94 and @pangtao22)

___
*[src/iiwa_plan_manager.h, line 26 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuF8CkCo-pDdrAxemx:-MYuL2NW2fOf3ExRqkOb:b-896fix) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/iiwa_plan_manager.h#L26)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

Is it possible to make it constant?
</blockquote></details>

Done.
___
*[src/plan_manager_system/iiwa_plan_manager_hardware_interface.h, line 23 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuFhQN1s0EqQi6f_iN:-MYuKz9c03CPzdmGYgjl:bs7wumx) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/plan_manager_system/iiwa_plan_manager_hardware_interface.h#L23)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

`IiwaPlanManagerHardwareInterface` doesn't seem to use config outside the constructor. Maybe we don't need to save it as a data member here? It's probably better for the config to be owned by `IiwaPlanManagerSystem`?
</blockquote></details>

Done. 
___
*[src/plan_manager_system/iiwa_plan_manager_system.h, line 48 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuHBon30l4O6lnThlN:-MYuKvqV5iTZBWUGnPxV:b1ljg7p) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/plan_manager_system/iiwa_plan_manager_system.h#L48)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

Is it possible to make it constant, similar to `IiwaPlanManager`?
</blockquote></details>

Done. The constructor arguments have also changed to const. 
___
*[src/plans/iiwa_plan_factory.cc, line 23 at r4](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-MYuHiR-8UolhxEIBa6M:-MYuL47J0UBqW7snLZfv:b-1pq4ew) ([raw file](https://github.com/robotlocomotion/robot-plan-runner/blob/d7eb6cd2e03cca89f1acc43bf93a53d8dcdf4f7c/src/plans/iiwa_plan_factory.cc#L23)):*
<details><summary><i>Previously, pangtao22 wrote…</i></summary><blockquote>

nit: this can be made a constant reference.
</blockquote></details>

Done, but I had to get rid of the next line to make it const.

**Review by pangtao22 — COMMENTED** [review]
:lgtm_strong: I pushed some minor changes about passing by value/reference. 

> *[Reviewable](https://reviewable.io/reviews/robotlocomotion/robot-plan-runner/18#-:-MYuU6V_-hi51Wr2ZmCL:b-ufyjld)* status: 11 of 27 files reviewed, 3 unresolved discussions (waiting on @hjsuh94 and @mpetersen94)

**Review by pangtao22 — APPROVED** [review]

**Pull request merged**